penrose-web/src/compiler/Style.ts

```diff
@@ -1,7 +1,2068 @@
-export const compileStyle = (
-  env: VarEnv,
-  subAST: SubProg,
-  styAST: StyProg
-) /* :State */ => {
-  // TODO: fill in the style compiler here
+import * as _ from "lodash";
+import {
+  insertExpr, insertGPI, findExpr, insertExprs, valueNumberToAutodiff, valueNumberToAutodiffConst, isPath, defaultLbfgsParams, initConstraintWeight
+} from "engine/EngineUtils";
+import { canvasXRange, shapedefs, findDef, ShapeDef, PropType, IPropModel, IShapeDef, Sampler } from "shapes/ShapeDef";
+import { randFloats } from "utils/Util";
+import { numOf, constOf } from "engine/Autodiff";
+import { compileDomain, compileSubstance, checkDomain, checkSubstance, parseSubstance, parseDomain } from "API";
+
+import nearley from "nearley";
+import styleGrammar from "parser/StyleParser";
+import { SubstanceEnv, LabelMap, checkPredicate, checkVar, checkExpr, subtypeOf } from "compiler/Substance";
+import { Result, ok, err, unsafelyUnwrap, isErr } from "utils/Error";
+
+import { Env, isDeclaredSubtype, checkTypeConstructor } from "./Domain";
+import { Map } from "immutable"; // Note: Domain maps are immutable!!!
+
+// TEST PROGRAMS
+// COMBAK: Make the frontend testing easier/more programmatic...
+
+// -> WORKING
+// import { domainStr, subStrSugared, subStrUnsugared, styStr } from "compiler/TestLA";
+// import { domainStr, subStrSugared, subStrUnsugared, styStr } from "compiler/TestSetTree";
+
+// -> NOT WORKING
+// import { domainStr, subStrSugared, subStrUnsugared, styStr } from "compiler/TestSetCircles";
+import { domainStr, subStrSugared, subStrUnsugared, styStr } from "compiler/TestDebug";
+
+const clone = require("rfdc")({ proto: false, circles: false });
+
+// TODO: Write pseudocode / code comments / tests for the Style compiler
+
+// Really it should be SubOut, not SubProg:
+
+// -- | 'SubOut' is the output of the Substance compiler, comprised of:
+// -- * Substance AST
+// -- * (Variable environment, Substance environment)
+// -- * A mapping from Substance ids to their coresponding labels
+// data SubOut =
+//   SubOut SubProg (VarEnv, SubEnv) LabelMap
+
+//#region consts
+const ANON_KEYWORD = "ANON";
+const LOCAL_KEYWORD = "$LOCAL";
+
+const LABEL_FIELD = "label";
+
+const UnknownTagError = new Error("unknown tag");
+
+//#endregion
+
+//#region utils
+
+// numbers from 0 to r-1 w/ increment of 1
+const numbers = (r: number): number[] => {
+  const l = 0;
+  if (l > r) { throw Error("invalid range"); }
+  const arr = [];
+  for (let i = l; i < r; i++) {
+    arr.push(i);
+  }
+  return arr;
+}
+
+export function numbered<A>(xs: A[]): [A, number][] {
+  if (!xs) throw Error("fail");
+  return _.zip(xs, numbers(xs.length)) as [A, number][]; // COMBAK: Don't know why typescript has problem with this
```

Comment:
Because `zip` defaults to `undefined` if a list is longer, which is different from Haskell. Try numbering things with just `map` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map). It gives you the index too if all you want is the sequential numbering of items.